### PR TITLE
Np 45363 aggregation facet

### PR DESCRIPTION
**Pull request by StigNorland**

### CODE REVIEW

**Review by StigNorland — COMMENTED** [review]
tok ca 40 minutter å gå igjennom

File: search-commons/src/main/java/no/unit/nva/search2/ImportCandidateQuery.java [hunk 1]
```diff
@@ -175,10 +205,23 @@ private MultiMatchQueryBuilder multiMatchQuery() {
     boolean isFirstPage() {
         return ZERO.equals(getValue(FROM).toString());
     }
+
+    @NotNull
+    public static String[] extractFields(String field) {
+        return ALL.equals(field) || isNull(field)
+            ? ASTERISK.split(COMMA)
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/ImportCandidateQuery.java:212]: extract method for readability?

File: search-commons/src/main/java/no/unit/nva/search2/ImportCandidateQuery.java [hunk 2]
```diff
@@ -175,10 +205,23 @@ private MultiMatchQueryBuilder multiMatchQuery() {
     boolean isFirstPage() {
         return ZERO.equals(getValue(FROM).toString());
     }
+
+    @NotNull
+    public static String[] extractFields(String field) {
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/ImportCandidateQuery.java:210]: add to name that you are removing keywords from search string?

File: search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java [hunk 3]
```diff
@@ -136,79 +153,115 @@ private BoolQueryBuilder boolQuery(UserSettingsClient userSettingsClient) {
...
+
+    @JacocoGenerated
+    Tuple<String, SortOrder> expandSortKeys(String... strings) {
+        var sortOrder = strings.length == 2 ? SortOrder.fromString(strings[1]) : SortOrder.ASC;
+        var fieldName = ResourceSort.fromSortKey(strings[0]).getFieldName();
+        return new Tuple<>(fieldName, sortOrder);
+    }
+
+    @NotNull
+    public static String[] extractFields(String field) {
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java:222]: same as other, but missing keyword filtering

File: search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java [hunk 4]
```diff
@@ -136,79 +153,115 @@ private BoolQueryBuilder boolQuery(UserSettingsClient userSettingsClient) {
...
         var value = getValue(SEARCH_ALL).toString();
         return QueryBuilders
             .multiMatchQuery(value, fields)
             .type(Type.CROSS_FIELDS)
             .operator(Operator.AND);
     }
-    
-    boolean isFirstPage() {
+
+    public static URI nextResultsBySortKey(
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java:200]: same method as another in PR?

File: search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java [hunk 5]
```diff
@@ -275,65 +324,65 @@ protected void validateSort() throws BadRequestException {
...
-        private String expandDate(String value) {
-            return value.length() == 4 ? value + "-01-01" : value;
+
+        public String expandYearToDate(String value) {
+            return value.length() == 4 ? value + JANUARY_FIRST : value;
         }
-        
-        private String expandFields(String value) {
+
+        protected String expandFields(String value) {
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/ResourceQuery.java:376]: can all these expandFields method be conjoined somehow?

File: search-commons/src/main/java/no/unit/nva/search2/enums/ResourceSort.java [hunk 6]
```diff
@@ -11,31 +12,32 @@
...
-    CATEGORY("entityDescription.reference.publicationInstance.type"),
+    CATEGORY(ENTITY_DESCRIPTION_REFERENCE_PUBLICATION_INSTANCE_TYPE),
+    INSTANCE_TYPE(ENTITY_DESCRIPTION_REFERENCE_PUBLICATION_INSTANCE_TYPE),
     CREATED_DATE("createdDate"),
     MODIFIED_DATE("modifiedDate"),
     PUBLISHED_DATE("publishedDate"),
-    TITLE("entityDescription.mainTitle"),
-    UNIT_ID("entityDescription.contributors.affiliation.id"),
-    USER("(?i)(user)|(owner)", "resourceOwner.owner");
+    TITLE("entityDescription.mainTitle.keyword"),
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/enums/ResourceSort.java:22]: title -> keyword?

File: search-commons/src/main/java/no/unit/nva/search2/enums/ResourceSort.java [hunk 7]
```diff
@@ -11,31 +12,32 @@
...
+    UNIT_ID("entityDescription.contributors.affiliations.id.keyword"),
+    USER("(?i)(user)|(owner)", "resourceOwner.owner.keyword");
```
Comment (torarnet) [inline search-commons/src/main/java/no/unit/nva/search2/enums/ResourceSort.java:24]: same. Use constants?

File: search-commons/src/test/java/no/unit/nva/search2/ImportCandidateClientTest.java [hunk 8]
```diff
@@ -1,187 +1,213 @@
...
+        @MethodSource("uriProvider")
+        void searchWithUriReturnsOpenSearchAwsResponse(URI uri) throws ApiGatewayException {
+            var query =
+                ImportCandidateQuery.builder()
+                    .fromQueryParameters(queryToMapEntries(uri))
+                    .withOpensearchUri(URI.create(container.getHttpHostAddress()))
+                    .withRequiredParameters(FROM, SIZE, SORT)
+                    .build();
+
+            var response = importCandidateClient.doSearch(query);
```
Comment (torarnet) [inline search-commons/src/test/java/no/unit/nva/search2/ImportCandidateClientTest.java:91]: swsResponse?

File: search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java [hunk 9]
```diff
@@ -1,206 +1,333 @@
...
-            public HttpRequest request() {
-                return null;
-            }
+            assertNotNull(pagedSearchResourceDto);
 
-            @Override
-            public Optional<HttpResponse<Object>> previousResponse() {
-                return Optional.empty();
-            }
+            pagedSearchResourceDto.aggregations()
```
Comment (torarnet) [inline search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java:118]: Create test for asserting aggregations on the whole set

File: search-commons/src/test/java/no/unit/nva/search2/ResourceQueryTest.java [hunk 10]
```diff
@@ -102,53 +101,47 @@ void uriParamsDateToResourceParams(URI uri) throws BadRequestException {
...
     static Stream<URI> uriProvider() {
         return Stream.of(
             URI.create("https://example.com/"),
             URI.create("https://example.com/?fields=category,title,created_date"),
             URI.create("https://example.com/?query=Muhammad+Yahya&fields=CONTRIBUTOR"),
             URI.create("https://example.com/?CONTRIBUTOR=https://api.dev.nva.aws.unit.no/cristin/person/1136254"),
-            URI.create("https://example.com/?CONTRIBUTOR_SHOULD="
-                + "https://api.dev.nva.aws.unit.no/cristin/person/1136254+"
-                + "https://api.dev.nva.aws.unit.no/cristin/person/1135555"),
+            //            URI.create("https://example.com/?CONTRIBUTOR_SHOULD="
```
Comment (torarnet) [inline search-commons/src/test/java/no/unit/nva/search2/ResourceQueryTest.java:139]: remove commented code?

**Review by torarnet — COMMENTED** [review]
See comments

File: search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java [hunk 11]
```diff
@@ -1,206 +1,347 @@
...
+            assertFalse(aggregations.isEmpty());
+            assertThat(aggregations.get("userAffiliation").size(), is(3));
+            assertThat(aggregations.get("contextType").size(), is(3));
+            assertThat(aggregations.get("instanceType").size(), is(3));
+            assertThat(aggregations.get("associatedArtifacts").size(), is(1));
+            assertThat(aggregations.get("associatedArtifacts").get(0).count(), is(20));
+            assertThat(aggregations.get("fundingSource").size(), is(2));
+            assertThat(aggregations.get("user").size(), is(3));
+            assertThat(aggregations.get("topLevelOrganization").size(), is(4));
+            assertThat(aggregations.get("topLevelOrganization").get(1).labels().get("nb"),
```
[the remark this answers is not in the document]
Comment (truhacevkir) [reply search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java:120]: Facet for users should contain user name as well:
        generateSimpleAggregation("user", "entityDescription.contributors.identity.name.keyword")
        
 It should be possible to make partiall name search

[the remark this answers is not in the document]
Comment (truhacevkir) [reply search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java:120]: Sjekk om vi faktisk trenger denne aggregeringen:
```

    "userAffiliation": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
            {
                "key": "https://api.sandbox.nva.aws.unit.no/cristin/organization/7452.0.0.0",
                "doc_count": 11
            },
            {
                "key": "https://api.sandbox.nva.aws.unit.no/cristin/organization/185.90.0.0",
                "doc_count": 7
            },
            {
                "key": "https://api.dev.nva.aws.unit.no/cristin/organization/20754.0.0.0",
                "doc_count": 2
            }
        ]
    }
```

[the remark this answers is not in the document]
Comment (truhacevkir) [reply search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java:120]: Usikker om vi trenger denne:

```
    "contextType": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
            {
                "key": "Report",
                "doc_count": 10
            },
            {
                "key": "UnconfirmedJournal",
                "doc_count": 9
            },
            {
                "key": "Anthology",
                "doc_count": 1
            }
        ]
    }
```

[the remark this answers is not in the document]
Comment (truhacevkir) [reply search-commons/src/test/java/no/unit/nva/search2/ResourceClientTest.java:120]: kanskje endre navnet på associatedArtifacts aggregeringen til file/files/hasFiles

=> slik at søke blir slik hasFile=true ?

instanceType aggregeringsnavn til type?

File: search-commons/src/test/java/no/unit/nva/search2/ResourceClientNoHitsTest.java [hunk 12]
```diff
@@ -75,19 +73,18 @@ void uriParamsToResourceParams(URI uri) throws ApiGatewayException {
...
-
     static Stream<URI> uriSortingProvider() {
         return Stream.of(
-            URI.create("https://example.com/?category=PhdThesis&sort=category&sortOrder=asc&sort=created_date&order"
+            URI.create("https://example.com/?INSTANCE_TYPE=PhdThesis&sort=INSTANCE_TYPE&sortOrder=asc&sort=created_date&order"
                        + "=desc"),
-            URI.create("https://example.com/?category=PhdThesis"),
-            URI.create("https://example.com/?category=PhdThesis&orderBy=category:asc,created_date:desc"),
-            URI.create("https://example.com/?category=PhdThesis&sort=category+asc&sort=created_date+desc"));
+            URI.create("https://example.com/?INSTANCE_TYPE=PhdThesis"),
```
Comment (truhacevkir) [inline search-commons/src/test/java/no/unit/nva/search2/ResourceClientNoHitsTest.java:80]: UriWrapper.fromUri("https://example.com/").addQueryParameter(INSTANCE_TYPE, value).getUri();

Da slipper du å rename manuelt i alle tester

File: search-commons/src/main/java/no/unit/nva/search2/enums/ResourceParameter.java [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        return result.size() == 1
+            ? result.stream().findFirst().get()
+            : INVALID;
+    }
+
+    private static boolean ignoreInvalidKey(ResourceParameter f) {
+        return f.ordinal() > IGNORE_PARAMETER_INDEX;
+    }
+
+    private static boolean isSearchField(ResourceParameter f) {
```
Comment (truhacevkir) [inline search-commons/src/main/java/no/unit/nva/search2/enums/ResourceParameter.java:275]: f -> resourceParameter

File: search-commons/src/main/java/no/unit/nva/search2/constant/ImportCandidate.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    public static final String TYPE_KEYWORD = "type.keyword";
+
+    public static final List<AbstractAggregationBuilder<? extends AbstractAggregationBuilder<?>>>
+        IMPORT_CANDIDATES_AGGREGATIONS = List.of(
+        generateSimpleAggregation(CANDIDATE_STATUS, STATUS_TYPE_KEYWORD),
+        generateSimpleAggregation(PUBLICATION_YEAR, PUBLICATION_YEAR_KEYWORD),
+        generateSimpleAggregation(INSTANCE_TYPE, PUBLICATION_INSTANCE_TYPE),
+        generateSimpleAggregation(COLLABORATION_TYPE, COLLABORATION_TYPE_KEYWORD),
+        generateSimpleAggregation(IMPORTED_BY_USER, IMPORT_STATUS_SET_BY_KEYWORD),
+        generateObjectLabelsAggregation("organization", "organizations"),
```
Comment (truhacevkir) [inline search-commons/src/main/java/no/unit/nva/search2/constant/ImportCandidate.java:45]: can be a constant

File: search-commons/src/main/java/no/unit/nva/search2/constant/Functions.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+    static TermsAggregationBuilder generateHasFileAggregation() {
+        var include = new IncludeExclude(PUBLISHED_FILE, "");
+        return AggregationBuilders
+            .terms(ASSOCIATED_ARTIFACTS)
+            .field(jsonPath(ASSOCIATED_ARTIFACTS, TYPE, KEYWORD))
+            .includeExclude(include)
+            .size(Defaults.DEFAULT_AGGREGATION_SIZE)
+            .subAggregation(
+                generateSimpleAggregation("public", jsonPath(ASSOCIATED_ARTIFACTS, ADMINSTRATIVE_AGREEMENT))
```
Comment (truhacevkir) [inline search-commons/src/main/java/no/unit/nva/search2/constant/Functions.java:77]: public can also be a constant

File: search-commons/src/main/java/no/unit/nva/search2/constant/Functions.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    static NestedAggregationBuilder generateLabelsAggregation(String jsonPath) {
+        var nestedAggregation = new NestedAggregationBuilder(LABELS, jsonPath(jsonPath, LABELS));
+        Stream.of(BOKMAAL_CODE, ENGLISH_CODE, NYNORSK_CODE, SAMI_CODE)
+            .map(code -> generateSimpleAggregation(code, jsonPath(jsonPath, LABELS, code, KEYWORD)))
+            .forEach(nestedAggregation::subAggregation);
+        return nestedAggregation;
+    }
+
+    static TermsAggregationBuilder generateHasFileAggregation() {
+        var include = new IncludeExclude(PUBLISHED_FILE, "");
```
Comment (truhacevkir) [inline search-commons/src/main/java/no/unit/nva/search2/constant/Functions.java:70]: StringUtil.EMPTY_STRING

File: search-commons/src/main/java/no/unit/nva/search2/common/QueryBuilderTools.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+                }
+                case MUST_NOT -> bq.mustNot(termsQuery);
+                case SHOULD -> bq.should(termsQuery);
+                default -> throw new IllegalArgumentException(OPERATOR_NOT_SUPPORTED);
+            }
+        });
+    }
+
+    public static QueryBuilder buildQuery(ParameterKey key, String value) {
+        final var values = value.replace(COMMA, " ");
```
Comment (truhacevkir) [inline search-commons/src/main/java/no/unit/nva/search2/common/QueryBuilderTools.java:50]: constant

**Review by truhacevkir — APPROVED** [review]

**Pull request merged**